File: Codes from the AI-assistants/Old Experiments/[CLBG] fannkuch-redux/Specialized Prompt - TBR Mode/FannkuchRedux-gpt-specialized-tbr.c
// fannkuch.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv) {
    int n = (argc > 1) ? atoi(argv[1]) : 7;
    int i, j, k;

    int *perm = malloc(n * sizeof(int));
    int *perm1 = malloc(n * sizeof(int));
    int *count = malloc(n * sizeof(int));
    int *temp = malloc(n * sizeof(int));
    if (!perm || !perm1 || !count || !temp) { perror("alloc"); return 1; }

    for (i = 0; i < n; ++i) perm1[i] = i + 1;
    for (i = 0; i < n; ++i) {
        perm[i] = perm1[i];
        count[i] = i;
    }

    int maxflips = 0;
    int checksum = 0;
    int sign = 1;

    while (1) {
        // copy perm -> temp for flipping
        memcpy(temp, perm, n * sizeof(int));

        // count flips
        int flips = 0;
        while (temp[0] != 1) {
            k = temp[0];
            // reverse prefix of length k
            for (i = 0, j = k - 1; i < j; ++i, --j) {
                int t = temp[i]; temp[i] = temp[j]; temp[j] = t;
            }
            flips++;
            if (flips > n) break; // safety
        }
        checksum += sign * flips;
        if (flips > maxflips) maxflips = flips;

        // generate next permutation using count (odometer)
        int idx = 1;
        while (idx < n) {
            if (count[idx] > 0) {
                // rotate first idx+1 elements left by 1
                int first = perm[0];
                for (i = 0; i < idx; ++i) perm[i] = perm[i + 1];
                perm[idx] = first;

                count[idx]--;
                sign = -sign;
                break;
            } else {
                count[idx] = idx;
                idx++;
            }
        }
        if (idx >= n) break; // done
    }

    printf("%d\nPfannkuchen(%d) = %d\n", checksum, n, maxflips);

    free(perm); free(perm1); free(count); free(temp);
    return 0;
}

File: Codes from the AI-assistants/Old Experiments/[CLBG] fannkuch-redux/Specialized Prompt - TBR Mode/FannkuchRedux-gpt-specialized-tbr.py
# fannkuch.py
import sys

def fannkuch(n):
    perm = list(range(1, n+1))
    count = list(range(n))
    maxflips = 0
    checksum = 0
    sign = 1

    while True:
        # count flips
        temp = perm[:]  # single allocation per permutation (O(n))
        flips = 0
        while temp[0] != 1:
            k = temp[0]
            temp[:k] = reversed(temp[:k])
            flips += 1
        checksum += sign * flips
        if flips > maxflips:
            maxflips = flips

        # next permutation using count[] as odometer
        i = 1
        while i < n:
            if count[i] > 0:
                # rotate left first i+1
                first = perm.pop(0)
                perm.append(first)
                count[i] -= 1
                sign = -sign
                break
            else:
                count[i] = i
                i += 1
        if i >= n:
            break
    return checksum, maxflips

if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 7
    checksum, maxflips = fannkuch(n)
    print(checksum)
    print(f"Pfannkuchen({n}) = {maxflips}")
